crates/sui-types/src/storage.rs

```diff
@@ -157,3 +163,329 @@ impl<S: ChildObjectResolver> ChildObjectResolver for &S {
+    }
+
+    fn insert_committee(&self, new_committee: Committee) {
+        WriteStore::insert_committee(*self, new_committee)
+    }
+
+    fn insert_transaction(&self, transaction: Transaction) {
+        WriteStore::insert_transaction(*self, transaction)
+    }
+
+    fn insert_transaction_effects(&self, transaction_effects: TransactionEffects) {
+        WriteStore::insert_transaction_effects(*self, transaction_effects)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct InMemoryStore {
+    highest_trusted_checkpoint: Option<(CheckpointSequenceNumber, CheckpointDigest)>,
+    highest_synced_checkpoint: Option<(CheckpointSequenceNumber, CheckpointDigest)>,
+    // highest_executed_checkpoint: Option<(CheckpointSequenceNumber, CheckpointDigest)>,
+    checkpoints: HashMap<CheckpointDigest, VerifiedCheckpoint>,
+    sequence_number_to_digest: HashMap<CheckpointSequenceNumber, CheckpointDigest>,
+    checkpoint_contents: HashMap<CheckpointContentsDigest, CheckpointContents>,
+    transactions: HashMap<TransactionDigest, Transaction>,
+    effects: HashMap<TransactionEffectsDigest, TransactionEffects>,
+
+    epoch_to_committee: Vec<Committee>,
+}
+
+impl InMemoryStore {
+    pub fn get_checkpoint_by_digest(
+        &self,
+        digest: &CheckpointDigest,
+    ) -> Option<&VerifiedCheckpoint> {
+        self.checkpoints.get(digest)
+    }
+
+    pub fn get_checkpoint_by_sequence_number(
+        &self,
+        sequence_number: CheckpointSequenceNumber,
+    ) -> Option<&VerifiedCheckpoint> {
+        self.sequence_number_to_digest
+            .get(&sequence_number)
+            .and_then(|digest| self.get_checkpoint_by_digest(digest))
+    }
+
+    pub fn get_highest_trusted_checkpoint(&self) -> Option<&VerifiedCheckpoint> {
+        self.highest_trusted_checkpoint
+            .as_ref()
+            .and_then(|(_, digest)| self.get_checkpoint_by_digest(digest))
+    }
+
+    pub fn get_highest_synced_checkpoint(&self) -> Option<&VerifiedCheckpoint> {
+        self.highest_synced_checkpoint
+            .as_ref()
+            .and_then(|(_, digest)| self.get_checkpoint_by_digest(digest))
+    }
+
+    pub fn get_checkpoint_contents(
+        &self,
+        digest: &CheckpointContentsDigest,
+    ) -> Option<&CheckpointContents> {
+        self.checkpoint_contents.get(digest)
+    }
+
+    pub fn insert_checkpoint_contents(&mut self, contents: CheckpointContents) {
+        self.checkpoint_contents.insert(contents.digest(), contents);
+    }
+
+    // Insert a verified checkpoint
+    pub fn insert_checkpoint(&mut self, checkpoint: VerifiedCheckpoint) {
+        let digest = checkpoint.digest();
+        let sequence_number = checkpoint.sequence_number();
+
+        if let Some(next_committee) = checkpoint.next_epoch_committee() {
+            let next_committee = next_committee.iter().cloned().collect();
+            let committee = Committee::new(checkpoint.epoch().saturating_add(1), next_committee)
```

Comment:
Is this function only called from local validator path? i.e. could it not be coming from the network?

Reply:
No it can definitely be called for a checkpoint we got from the network (eg the fullnode case)